### ISSUE TITLE
AllGatherFunc方法backward最后要对梯度再乘以len(grad_list)也就是world_size?

### ISSUE THREAD

**Issue by gobigrassland** [Tencent/TFace#34]
AllGatherFunc方法backward最后要对梯度再乘以len(grad_list)也就是world_size? 这个没有想明白。还望大牛给解释一下。
之前关于Allgather这块是如何进行梯度传递没有了解，这几天整理了一下。主要包括2个问题：
（1）为什么要对grad_out求和？
（2）为什么要乘以world_size?
第（1）点我查看了一些资料并推导了一下公式，差不多理解了。下面写出来，大家看看我的理解是否正确哈。第（2）还是没有理解。

下面用示意图来描述大规模人脸分类的过程，结合下面示意图和公式表达来理解。
B: batch size, d: feature dimension, K: gpu number, C: class number, $c_j$: class number of j-th gpu
（1）<img src="https://render.githubusercontent.com/render/math?math=F_j \in R^{B*d}">：第j块GPU上特征
（2）<img src="https://render.githubusercontent.com/render/math?math=F_{total} = torch.cat((F_0, F_1, ^, F_{K-1} )) \in R^{KB*d}">：表示所有的K个GPU上特征合并在一起
（3）<img src="https://render.githubusercontent.com/render/math?math=W_j \in R^{d*c_j}">：表示所有的K个GPU上特征合并在一起
（4）<img src="https://render.githubusercontent.com/render/math?math=logit_j=F_{total}W_j \in R^{KB*c_j}">：这里简化分类层为常规线性变换。（下面的公式中$y_j$就表示$logit_j$）
（5）<img src="https://render.githubusercontent.com/render/math?math=\frac {\partial L_j}{\partial F_{total}} = \frac{\partial L_j}{\partial y_j}* \frac{\partial y_j}{\partial F_{total}}=\frac{\partial L_j}{\partial y_j}*W_j^T">

   可以看出每块GPU上产生的对全体特征向量的梯度维度都是一样（这个是肯定的），每块GPU上产生梯度是通过上述链式法则得到的，得到梯度的公式中，分两个部分相乘，一个是对logit值的导数，一个是当前卡上局部分类权重W的导数。对于每块卡而言这两部分都不一样。也就是每块gpu都对全体特征向量$F_{total}$都产生梯度。总的loss是各个GPU上loss先求和再归约，因此在求对logit梯度时，也除以了总的样本数量（KB），然后对全体特征向量$F_{total}$在allgather层要进行相加。
<img src="https://render.githubusercontent.com/render/math?math=\frac{\partial L}{\partial F_{total}}=\frac{1}{KB}\sum _{j=0}^{j=K-1}\frac {\partial L_j}{\partial F_{total}} =\frac{1}{KB}\sum _{j=0}^{j=K-1}\frac{\partial L_j}{\partial y_j}*W_j^T=\sum _{j=0}^{j=K-1}\frac{1}{KB}\frac{\partial L_j}{\partial y_j}*W_j^T">
上面是对第（1）点的回答。大家看看我的解释是否正确。
![](https://img2020.cnblogs.com/blog/1029859/202109/1029859-20210929203250229-1981508692.png)

（2）可是至此，还不明白上述代码为什么要乘以GPU的数量,对应代码为：```grad_out *= len(grad_list) ```。
在https://amsword.medium.com/gradient-backpropagation-with-torch-distributed-all-gather-9f3941a381f8 也提到了要乘以world_size,但是还是没有明白。

**Comment by wjxzju** [comment]
这里主要是为了distfc和localfc的对齐，grad_out乘以GPU卡数后，对于每张卡上的F_j来说刚好和localfc的时候是等价的

**Comment by SCZwangxiao** [comment]
因为DistributedDataParallel会自动平均所有GPU上的梯度，所以要补偿一个world_size的倍数。
或者你可以这么理解：没有all_gather他就是数据并行，DDP自动平均梯度；有all_gather就是模型并行，DDP自动平均的做法是多余的。

**Comment by Kipsora** [comment]
@gobigrassland 为什么最后是除以了 `KB`？看代码是先 `all_reduce` 然后再取 `mean`，但 `all_reduce` 是用的 `sum`，这不应该只除以了 `B` 嘛（最后的对 `loss` 取 `mean`，这里 `loss` 的 `shape` 应该是 `[B, 1]`）：

TFace 的相关代码：

https://github.com/Tencent/TFace/blob/9004edba43e27770b8915aa3d6c42b197e4886bb/recognition/torchkit/loss/dist_softmax.py#L34-L37

Insightface 的[代码](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/partial_fc.py#L463-L467)也一样。

如果这里乘了 `world_size`，那不就变成了 `1 / B / K * K = 1 / B` 嘛（除以 `K` 是因为 DDP 带来的），但我们本来要的难道不就是 `1 / KB` 而非 `1 / B`？

**Comment by anxiangsir** [comment]
> @gobigrassland 为什么最后是除以了 `KB`？看代码是先 `all_reduce` 然后再取 `mean`，但 `all_reduce` 是用的 `sum`，这不应该只除以了 `B` 嘛（最后的对 `loss` 取 `mean`，这里 `loss` 的 `shape` 应该是 `[B, 1]`）：
> 
> TFace 的相关代码：
> 
> https://github.com/Tencent/TFace/blob/9004edba43e27770b8915aa3d6c42b197e4886bb/recognition/torchkit/loss/dist_softmax.py#L34-L37
> 
> Insightface 的[代码](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/partial_fc.py#L463-L467)也一样。
> 
> 如果这里乘了 `world_size`，那不就变成了 `1 / B / K * K = 1 / B` 嘛（除以 `K` 是因为 DDP 带来的），但我们本来要的难道不就是 `1 / (KB)` 而非 `1 / B`？

这里的loss是为了打印的时候用的，它不会参与到反向传播，影响到梯度大小的是softmax中backward里对梯度的缩放
https://github.com/Tencent/TFace/blob/master/recognition/torchkit/loss/dist_softmax.py#L52
乘 world_size 之前，梯度为1 / B / K，乘完之后为 1 / B，DDP继续除K得到 1 / B / K
